utils/mgmt_system.py

```diff
@@ -71,6 +77,32 @@ def get_search_filter_spec(self, *args, **kwargs):
 
         return pfs
 
+    def __getattribute__(self, attr):
+        # fetch the attribute using parent class to avoid recursion
+        res = super(_PsphereClient, self).__getattribute__(attr)
+        # any callable (except login) is login protected
+        if callable(res) and attr not in ['login']:
+            # fetch the cached, login-protected version
+            try:
+                return self._cached_protect[res]
+            # or create it and save to cache
+            except KeyError:
+                # simple try-except wrapper that tries to log in on failure
+                def login_protect(client, o):
```

Comment:
This function is going to get created every time `__getattribute__` called. It needs to be defined somewhere that will only run once, such as the module scope or the class scope (probably as a static method).